src/appleseed-max-impl/osloutputselectormap/osloutputselector.cpp

```diff
@@ -0,0 +1,604 @@
+
+    enum { ParamBlockIdOutputSelector };
+    enum { ParamBlockRefOutputSelector };
+
+    enum ParamId
+    {
+        ParamIdSourceMap = 0,
+        ParamIdOutputIndex = 1,
+    };
+
+    enum TexmapId
+    {
+        TexmapIdSourceMap = 0,
+        TexmapCount                 // keep last
+    };
+
+    const MSTR g_texmap_slot_names[TexmapCount] =
+    {
+        L"SourceMap",
+    };
+
+    const ParamId g_texmap_id_to_param_id[TexmapCount] =
+    {
+        ParamIdSourceMap
+    };
+
+    class OutputIndexAccessor
+      : public PBAccessor
+    {
+      public:
+        void Set(PB2Value& v, ReferenceMaker* owner, ParamID id, int tabIndex, TimeValue t) override
+        {
+            IParamBlock2* pblock = static_cast<OSLOutputSelector*>(owner)->GetParamBlock(0);
+            if (pblock == nullptr)
+                return;
+
+            IParamMap2* map = pblock->GetMap();
+            if (map == nullptr)
+                return;
+
+            switch (id)
+            {
+                case ParamIdOutputIndex:
+                {
+                    Texmap* texmap = nullptr;
+                    pblock->GetValue(ParamIdSourceMap, t, texmap, FOREVER);
+
+                    if (texmap != nullptr && is_osl_texture(texmap))
+                    {
+                        auto output_names = static_cast<OSLTexture*>(texmap)->get_output_names();
+
+                        int output_index = v.i;
+                        if (!output_names.empty())
+                        {
+                            HWND dlg_hwnd = 0;
+                            dlg_hwnd = map->GetHWnd();
+                            if (dlg_hwnd != 0)
+                            {
+                                SendMessage(
+                                    GetDlgItem(dlg_hwnd, IDC_COMBO_OUTPUT_NAME),
+                                    CB_SETCURSEL,
+                                    (output_index - 1) % output_names.size(),
+                                    0);
+                            }
+                        }
+                    }
+                }
+            }
+        }
+    };
+}
+
+OSLOutputSelectorClassDesc g_appleseed_outputselector_classdesc;
+OutputIndexAccessor g_output_index_accessor;
+
+//
+// OSLOutputSelector class implementation.
+//
+
+namespace
+{
+    ParamBlockDesc2 g_block_desc(
+        // --- Required arguments ---
+        ParamBlockIdOutputSelector,                 // parameter block's ID
+        L"appleseedOuputSelectorParams",            // internal parameter block's name
+        0,                                          // ID of the localized name string
+        &g_appleseed_outputselector_classdesc,      // class descriptor
+        P_AUTO_CONSTRUCT + P_AUTO_UI,               // block flags
+
+         // --- P_AUTO_CONSTRUCT arguments ---
+        ParamBlockRefOutputSelector,                // parameter block's reference number
+
+        // --- P_AUTO_UI arguments for Parameters rollup ---
```

Comment:
Add a blank line after this comment since it applies not only to the next block, but to all following blocks.